lib/grandpa/finalisation.go

```diff
@@ -0,0 +1,467 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ChainSafe/gossamer/dot/telemetry"
+)
+
+var (
+	errServicesStopFailed       = errors.New("services stop failed")
+	errvotingRoundHandlerFailed = errors.New("voting round ephemeral failed")
+	errfinalisationEngineFailed = errors.New("finalisation engine ephemeral failed")
+)
+
+type ephemeralService interface {
+	Run() error
+	Stop() error
+}
+
+type finalisationHandler struct {
+	servicesLock       sync.Mutex
+	finalisationEngine ephemeralService
+	votingRound        ephemeralService
+
+	// newServices is a constructor function which takes care to instantiate
+	// and return the services needed to finalize a round, those services
+	// are ephemeral services with a lifetime of a round
+	newServices   func() (engine, voting ephemeralService)
+	initiateRound func() error
+
+	stopCh      chan struct{}
+	handlerDone chan struct{}
+}
+
+func newFinalisationHandler(service *Service) *finalisationHandler {
+	return &finalisationHandler{
+		newServices: func() (engine, voting ephemeralService) {
+			finalisationEngine := newfinalisationEngine(service)
+			votingRound := newvotingRoundHandler(service, finalisationEngine.actionCh)
+			return finalisationEngine, votingRound
+		},
+		initiateRound: service.initiateRound,
+		stopCh:        make(chan struct{}),
+		handlerDone:   make(chan struct{}),
+	}
+}
+
+func (fh *finalisationHandler) Start() (<-chan error, error) {
+	errorCh := make(chan error)
+	go fh.run(errorCh)
+
+	return errorCh, nil
+}
+
+func (fh *finalisationHandler) run(errorCh chan<- error) {
+	defer func() {
+		close(errorCh)
+		close(fh.handlerDone)
+	}()
+
+	for {
+		select {
+		case <-fh.stopCh:
+			return
+		default:
+		}
+
+		err := fh.initiateRound()
+		if err != nil {
+			errorCh <- fmt.Errorf("initiating round: %w", err)
+			return
+		}
+
+		err = fh.runEphemeralServices()
+		if err != nil {
+			errorCh <- fmt.Errorf("running ephemeral services: %w", err)
+			return
+		}
+	}
+}
+
+func (fh *finalisationHandler) stop() (err error) {
+	fh.servicesLock.Lock()
```

Comment:
What is actually being locked here? Would other processes be trying to read/write to the ephemeral services?

Reply:
Since the services are writen in the `fh.run()` goroutine but we read those services at `fh.stop()` so we have this `servicesLock` to prevent rece conditions.

Reply:
ah okay I see, thank you!